File: Strings/LongestCommonSubstring.java
public class LongestCommonSubstring{
public static String longest(String S,String T){
  int[][] table=new int[S.length()+1][T.length()+1];
  int maxLen=0;
  String result=null;
  for(int i=1;i<=S.length();i++){
    for(int j=1;j<=T.length();j++){
      if(S.charAt(i-1)==T.charAt(j-1)){
        table[i][j]=table[i-1][j-1]+1;
        if(table[i][j]>maxLen){
          maxLen=table[i][j];
          result=S.substring(i-maxLen,i);
        }
      }
    }
  }
  System.out.println(maxLen);
  return result;
}
public static void main(String[] args){
  System.out.println(longest("ababc","babac"));
    System.out.println(longest("abac","babac"));

}
}

File: Strings/longestCommonSubstring.py
def lcs(s1,s2):
    table=[[0]*(1+len(s2)) for x in xrange(1+len(s1))]
    maxLen=0
    res=""
    for i in range (1,len(s1)+1):
        for j in range (1,len(s2)+1):
            if(s1[i-1]==s2[j-1]):
                table[i][j]=table[i-1][j-1]+1
                if table[i][j]>maxLen:
                    maxLen=table[i][j]
                    res=s1[i-maxLen:i]
    print maxLen
    return res


print lcs("abac","babac")

print lcs("ababc","babac")
